src/scala/microsoft-spark-2.4.x/src/main/scala/org/apache/spark/api/dotnet/CallbackConnection.scala

```diff
@@ -0,0 +1,143 @@
+/*
+ * Licensed to the .NET Foundation under one or more agreements.
+ * The .NET Foundation licenses this file to you under the MIT license.
+ * See the LICENSE file in the project root for more information.
+ */
+
+package org.apache.spark.api.dotnet
+
+import java.io.{Closeable, DataInputStream, DataOutputStream}
+import java.net.Socket
+
+import org.apache.spark.internal.Logging
+
+/**
+ * CallbackConnection is used to process the callback communication
+ * between the JVM and Dotnet. It uses a TCP socket to communicate with
+ * the Dotnet CallbackServer and the socket is expected to be reused.
+ * @param address The address of the Dotnet CallbackServer
+ * @param port The port of the Dotnet CallbackServer
+ */
+class CallbackConnection(address: String, port: Int) extends Logging {
+  private[this] val socket = new Socket(address, port)
+  private[this] val inputStream = new DataInputStream(socket.getInputStream)
+  private[this] val outputStream = new DataOutputStream(socket.getOutputStream)
+
+  def send[T](
+    writeBody: DataOutputStream => Unit,
+    readBody: Option[DataInputStream => T]): CallbackResponse[T] = {
+    logInfo("Calling callback...")
```

Comment:
Can you paste overall logs? I think it will be useful to log callback id somewhere for debugging.

Reply:
At the moment logs look like:
```
	Line 38: 20/06/21 00:39:29 INFO DotnetBackendHandler: Connecting to a callback server at 127.0.0.1:61159
	Line 64: 20/06/21 00:39:33 INFO CallbackConnection: Calling callback 1 ...
	Line 64: 20/06/21 00:39:33 INFO CallbackConnection: Calling callback 1 ...
	Line 65: 20/06/21 00:39:33 INFO CallbackConnection: Signaling END_OF_STREAM.
	Line 117: 20/06/21 00:39:35 INFO CallbackConnection: Received END_OF_STREAM signal. Calling callback 1 successful.
	Line 117: 20/06/21 00:39:35 INFO CallbackConnection: Received END_OF_STREAM signal. Calling callback 1 successful.
	Line 196: 20/06/21 00:39:36 INFO CallbackConnection: Calling callback 1 ...
	Line 196: 20/06/21 00:39:36 INFO CallbackConnection: Calling callback 1 ...
	Line 197: 20/06/21 00:39:36 INFO CallbackConnection: Signaling END_OF_STREAM.
	Line 239: 20/06/21 00:39:37 INFO CallbackConnection: Received END_OF_STREAM signal. Calling callback 1 successful.
	Line 239: 20/06/21 00:39:37 INFO CallbackConnection: Received END_OF_STREAM signal. Calling callback 1 successful.
	Line 566: 20/06/21 00:39:40 INFO DotnetBackendHandler: Requesting to close callback client
	Line 567: 20/06/21 00:39:40 INFO CallbackClient: Shutting down.
```

Reply:
Why only one `Signaling END_OF_STREAM.`?

Reply:
Also, anything from C# side?

Reply:
there are 2 `Signaling END_OF_STREAM.`

Reply:
C# side looks like:

```
[2020-06-21T07:39:29.9722001Z] [UDON] [Info] [CallbackServer] Starting CallbackServer.
[2020-06-21T07:39:29.9860752Z] [UDON] [Info] [CallbackServer] Started CallbackServer on 127.0.0.1:61159
[2020-06-21T07:39:33.2653577Z] [UDON] [Info] [CallbackConnection] [1] Connected with RemoteEndPoint: 127.0.0.1:61160
[2020-06-21T07:39:33.2722132Z] [UDON] [Info] [CallbackServer] Pool snapshot: [NumThreads:1], [NumConnections:1]
[2020-06-21T07:39:33.2751166Z] [UDON] [Info] [CallbackConnection] [1] Received request for callback id: 1, callback handler: Microsoft.Spark.Interop.Ipc.ForeachBatchCallbackHandler
[2020-06-21T07:39:35.3845318Z] [UDON] [Debug] [CallbackConnection] [1] Received END_OF_STREAM signal.
[2020-06-21T07:39:36.6964764Z] [UDON] [Info] [CallbackConnection] [1] Received request for callback id: 1, callback handler: Microsoft.Spark.Interop.Ipc.ForeachBatchCallbackHandler
[2020-06-21T07:39:37.6173838Z] [UDON] [Debug] [CallbackConnection] [1] Received END_OF_STREAM signal.
```

Reply:
Ah ok, yea that was my confusion.

Can we change `Calling callback 1 ...` to `Calling callback [callback id = 1]...` or something like that?

Could you add these to PR description as well? This is very useful.